aws/resource_aws_security_group_attachment.go

```diff
@@ -0,0 +1,173 @@
+		return err
+	}
+
+	return addSGToENI(conn, sgID, dniResp.NetworkInterfaces[0])
+}
+
+func fetchNetworkInterface(conn *ec2.EC2, ifaceID string) (*ec2.NetworkInterface, error) {
+	dniParams := &ec2.DescribeNetworkInterfacesInput{
+		NetworkInterfaceIds: aws.StringSlice([]string{ifaceID}),
+	}
+
+	dniResp, err := conn.DescribeNetworkInterfaces(dniParams)
+	if err != nil {
+		return nil, err
+	}
+	return dniResp.NetworkInterfaces[0], nil
+}
+
+func addSGToENI(conn *ec2.EC2, sgID string, iface *ec2.NetworkInterface) error {
+	if sgExistsInENI(sgID, iface) {
+		return fmt.Errorf("security group %s already attached to interface ID %s", sgID, *iface.NetworkInterfaceId)
+	}
+	var groupIDs []string
+	for _, v := range iface.Groups {
+		groupIDs = append(groupIDs, *v.GroupId)
+	}
+	groupIDs = append(groupIDs, sgID)
+	params := &ec2.ModifyNetworkInterfaceAttributeInput{
+		NetworkInterfaceId: iface.NetworkInterfaceId,
+		Groups:             aws.StringSlice(groupIDs),
+	}
+
+	_, err := conn.ModifyNetworkInterfaceAttribute(params)
+	return err
+}
+
+func sgExistsInENI(sgID string, iface *ec2.NetworkInterface) bool {
+	for _, v := range iface.Groups {
+		if *v.GroupId == sgID {
+			return true
+		}
+	}
+	return false
+}
+
+func resourceAwsSecurityGroupAttachmentRead(d *schema.ResourceData, meta interface{}) error {
+	return refreshSecurityGroupWithInterface(d, meta)
+}
+
+func refreshSecurityGroupWithInterface(d *schema.ResourceData, meta interface{}) error {
+	sgID := d.Get("security_group_id").(string)
+	interfaceID := d.Get("network_interface_id").(string)
+
+	log.Printf("[INFO] Checking association of security group %s to network interface ID %s", sgID, interfaceID)
+
+	conn := meta.(*AWSClient).ec2conn
+
+	iface, err := fetchNetworkInterface(conn, interfaceID)
+	if err != nil {
+		return err
+	}
+
+	if sgExistsInENI(sgID, iface) {
+		d.SetId(fmt.Sprintf("%s_%s", sgID, interfaceID))
+	} else {
+		// The assocation does not exist when it should, taint this resource.
+		log.Printf("[WARN] Security group %s not associated with network interface ID %s, tainting", sgID, interfaceID)
+		d.SetId("")
+	}
+	return nil
+}
+
+func resourceAwsSecurityGroupAttachmentDelete(d *schema.ResourceData, meta interface{}) error {
+	if err := detachSecurityGroupFromInterface(d, meta); err != nil {
+		return err
+	}
+
+	d.SetId("")
+	return nil
+}
+
+func detachSecurityGroupFromInterface(d *schema.ResourceData, meta interface{}) error {
+	sgID := d.Get("security_group_id").(string)
+	interfaceID := d.Get("network_interface_id").(string)
+
+	log.Printf("[INFO] Removing security group %s from instance ID %s", sgID, interfaceID)
+
+	conn := meta.(*AWSClient).ec2conn
+
+	iface, err := fetchNetworkInterface(conn, interfaceID)
+	if err != nil {
+		return err
+	}
+
+	return delSGFromENI(conn, sgID, iface)
+}
+
+func delSGFromENI(conn *ec2.EC2, sgID string, iface *ec2.NetworkInterface) error {
+	old := iface.Groups
+	var new []*string
+	for _, v := range iface.Groups {
```

Comment:
Ack - good catch :fearful:

Do we have a provider-level mutex for this kind of thing?